File: client/src/service/productService.ts
import axiosInstance from "../config/axiosConfig";
import { ProductInterface } from "../util/types";

export class ProductService {
  static getAllProducts = async (
    page: number,
    subCategoryId?: number,
    size = 5,
  ) => {
    try {
      const response = await axiosInstance.get("/products", {
        params: { subCategoryId, page, size },
      });
      return response.data;
    } catch (error) {
      console.error("Get all products error", error);
      throw error;
    }
  };

  static getProductById = async (id: number): Promise<ProductInterface> => {
    try {
      const response = await axiosInstance.get<ProductInterface>(
        `/products/${id}`,
      );
      return response.data;
    } catch (error) {
      console.error("Get product error", error);
      throw error;
    }
  };

  static rateProduct = async (products: ProductInterface[]) => {
    try {
      const response = await axiosInstance.put("/products/rate", products);
      return response.data;
    } catch (error) {
      console.error("Rate products error", error);
      throw error;
    }
  };
}

File: server/src/main/java/com/clothify/server/service/ProductService.java
package com.clothify.server.service;

import com.clothify.server.entity.Product;
import org.springframework.data.domain.Page;

import java.util.List;
import java.util.Optional;

public interface ProductService {
    Page<Product> getAll(Long subCategoryId, int page, int size);

    Optional<Product> getById(long id);

    Product save(Product product);

    void delete(long id);

    void rateProduct(List<Product> products);

}
